cpp/src/utils/metrics/metric.hh

```diff
@@ -0,0 +1,50 @@
+#ifndef TS_CPP_UTILS_METRICS_METRIC_HH_
+#define TS_CPP_UTILS_METRICS_METRIC_HH_
+
+#include <string>
+#include <vector>
+
+#include "src/utils/metrics/dimension.hh"
+
+namespace torchserve {
+    enum MetricType {
+        COUNTER,
+        GAUGE,
+        HISTOGRAM
+    };
+
+    class MetricValue {
+        public:
+        MetricValue(const double& value, const uint64_t& timestamp, const std::string& request_id);
+        double GetValue() const;
+        uint64_t GetTimestamp() const;
+        const std::string GetRequestId() const;
+
+        private:
+        const double value;
+        const uint64_t timestamp;
+        const std::string request_id;
+    };
+
+    class Metric {
+        public:
+        Metric(const std::string& name, const std::string& unit,
+               const MetricType& type, const std::vector<Dimension>& dimensions);
+        const std::string GetName() const;
+        const std::string GetUnit() const;
+        MetricType GetType() const;
+        const std::vector<Dimension> GetDimensions() const;
+        const std::vector<MetricValue> GetValues() const;
+        void Update(const double& value, const std::string& request_id = "");
+        void Reset();
+
+        private:
+        const std::string name;
+        const std::string unit;
+        const MetricType type;
+        const std::vector<Dimension> dimensions;
+        std::vector<MetricValue> values;
```

Comment:
The data structure can be simplified and also prepared for sample data point fast updating, eg hashmap.

Reply:
Agree that it may be more efficient to use a hashmap to store the metric values, but a couple of quick questions on the suggested data structure `ConcurrentHashMap<Vector, Sample>` to be used inside the `Metric` object:

1. Since `Vector` is mutable, is it valid to use it as a key in a hashmap?
2. If we plan to store the `Sample` value associated with each possible combination of dimension values in a single `Metric` object, how would we support the `Metric` API `void Update(value);`? Everytime `Update` is called, the caller will need to supply the dimension values as well otherwise we would not know which `Sample` to update. Instead, if we store the value associated with a single metric with specific combination of dimension values in a single `Metric` object, wouldn't that make the implementation simpler to support the `MetricCache` APIs `Metric addMetric(metricName, unit, dimensions, metricType);`, `Metric getMetric(metricName, dimensions);` and also the `Metric` API `void Update(value);`? Although the tradeoff here is that we will use more space by duplicating metric attributes across `Metric` objects.